app/javascript/hyacinth_ui_v1/components/projects/permissions/PermissionsEditor.js

```diff
@@ -0,0 +1,300 @@
+  const [currentAddUserSelection, setCurrentAddUserSelection] = useState(null);
+
+  const { loading: actionsLoading, error: actionsError, data: actionsData } = useQuery(
+    getProjectPermissionActionsQuery, { variables: { stringKey: projectStringKey } },
+  );
+  const { loading: usersLoading, error: usersError } = useQuery(
+    getUsersQuery, {
+      onCompleted: (data) => {
+        setAllUsers(data.users);
+      },
+    },
+  );
+  const { loading: permissionsLoading, error: permissionsError } = useQuery(
+    getProjectPermissionsQuery, {
+      variables: { stringKey: projectStringKey },
+      onCompleted: (data) => {
+        setProjectPermissions(
+          collectionSortBy(data.projectPermissionsForProject, [perm => perm.user.sortName])
+        );
+      },
+    },
+  );
+
+  const [updateProjectPermissions, { error: updateProjectPermissionsError }] = useMutation(
+    updateProjectPermissionsMutation,
+  );
+
+  if (actionsLoading || permissionsLoading || usersLoading) return (<></>);
+  if (actionsError || permissionsError || usersError) {
+    return (
+      <GraphQLErrors errors={
+        actionsError || permissionsError || usersError || updateProjectPermissionsError
+      } />
+    );
+  }
+
+  const actionToDisplayLabel = (action) => {
+    return action.split('_').map(part => part.charAt(0).toUpperCase() + part.slice(1)).join(' ');
+  };
+
+  const updatePermissionsData = (userId, action, enabled) => {
+    setPermissionsChanged(true);
+
+    const { projectPermissionActions: {
+      actions: allActions, manageAction, actionsDisallowedForAggregatorProjects },
+    } = actionsData;
+
+    setProjectPermissions(
+      produce(projectPermissions, (draft) => {
+        const { actions: currentActionsForUser, project } = draft.find(
+          projectPermission => projectPermission.user.id === userId,
+        );
+        if (enabled) {
+          if (action === manageAction) {
+            // disable any current actions
+            currentActionsForUser.splice(0, currentActionsForUser.length);
+            // enable all allowed actions
+            if (project.isPrimary) {
+              currentActionsForUser.push(...allActions);
+            } else {
+              currentActionsForUser.push(...allActions.filter(
+                act => !actionsDisallowedForAggregatorProjects.includes(act)
+              ));
+            }
+          } else {
+            // enable single action
+            currentActionsForUser.push(action);
+          }
+        } else {
+          // disable single action
+          currentActionsForUser.splice(currentActionsForUser.indexOf(action), 1);
+        }
+      }),
+    );
+  };
+
+  const removeUser = (userId) => {
+    setPermissionsChanged(true);
+    setProjectPermissions(
+      produce(projectPermissions, (draft) => {
+        arrRemove(draft, (projectPermission) => {
+          return projectPermission.user.id === userId;
+        });
+      }),
+    );
+    setRemovedUserIds(
```

Comment:
```suggestion
    setRemovedUserIds(userIds => [...userIds, userId])
```

Reply:
Same question here as the earlier `setProjectPermissions`.  If re-using object elements from the previous array is fine when creating the new array, then I'm good with this suggested change.

Reply:
I think since they are just identifier values it should be fine. if they were objects that would be different, right?

Reply:
Yeah, true.  All identifiers should be fine, so I'll go the non-immer route, thanks!

The `removedUserIds` variable is a Set, so I went with:
`setRemovedUserIds(new Set([...removedUserIds, userId]));`

Just tested it out and it's working.  I ended up not using an arrow function to be more compact, but it's cool to know that you can supply an arrow function to a useState hook setter method.  So the hook automatically passes the current state val as the argument to whatever method you supply?